op-supervisor/supervisor/backend/db/logs/db.go

```diff
@@ -39,9 +30,168 @@ type Metrics interface {
+	if !l.doneBlock {
+		panic("logContext in invalid state, cannot retrieve block ID")
+	}
+	return l.blockHash
+}
+
+// TODO other getters, with checks if doneBlock and doneLog
+
+func (l *logContext) applyEntry(entry entrydb.Entry) {
+	if len(l.out) != 0 {
+		panic("can only apply without appending if the state is still empty")
+	}
+
+	// TODO decode entry
+
+	// TODO apply to state
+
+	l.lastType = entry.Type()
+	l.entryIndex += 1
+}
+
+func (l *logContext) appendEntry(obj EntryObj) {
+	entry := obj.encode()
+	l.out = append(l.out, entry)
+	l.lastType = entry.Type()
+	l.entryIndex += 1
+}
+
+func (l *logContext) infer() error {
+	if (l.entryIndex+1)%searchCheckpointFrequency == 0 {
+		l.appendEntry(newSearchCheckpoint(l.blockNum, l.logIdx, l.timestamp))
+		return nil
+	}
+	startLog := func() error {
+		if l.doneLog {
+			return io.EOF
+		}
+		evt := newInitiatingEvent(l.logHash, l.execMsg != nil)
+		l.appendEntry(evt)
+		if l.execMsg == nil {
+			l.doneLog = true
+		}
+		return nil
+	}
+	switch l.lastType {
+	case entrydb.TypeSearchCheckpoint:
+		l.appendEntry(newCanonicalHash(l.blockHash))
+		return nil
+	case entrydb.TypeCanonicalHash:
+		l.doneBlock = true
+		return startLog()
+	case entrydb.TypeInitiatingEvent:
+		if l.execMsg != nil {
+			link, err := newExecutingLink(*l.execMsg)
+			if err != nil {
+				return fmt.Errorf("failed to create executing link: %w", err)
+			}
+			l.appendEntry(link)
+			return nil
+		} else {
+			return startLog()
+		}
+	case entrydb.TypeExecutingLink:
+		l.appendEntry(newExecutingCheck(l.execMsg.Hash))
+		l.doneLog = true
+		return nil
+	case entrydb.TypeExecutingCheck:
+		return startLog()
+	case entrydb.TypeStartBlock:
+		l.appendEntry(newCanonicalHash(l.blockHash))
+		return nil
+	default:
+		return fmt.Errorf("unrecognized entry type: %s", l.lastType)
+	}
+}
+
+func (l *logContext) inferFull() error {
+	for {
+		err := l.infer()
+		if err == nil {
+			continue
+		}
+		if err == io.EOF { // wrapped io.EOF does not count.
+			return nil
+		} else {
+			return err
+		}
+	}
+}
+
+func (l *logContext) applyBlock(parent common.Hash, upd eth.BlockID, timestamp uint64) error {
+	if err := l.inferFull(); err != nil {
+		return err
+	}
+	if l.blockHash != types.TruncateHash(parent) {
+		return fmt.Errorf("cannot apply block %s (parent %s) on top of %s", upd, parent, l.blockHash)
+	}
+	if l.blockNum+1 != upd.Number {
+		return fmt.Errorf("cannot apply block %d on top of %d", upd.Number, l.blockNum)
+	}
+	if l.timestamp > timestamp {
+		return fmt.Errorf("block timestamp %d must be equal or larger than current timestamp %d", timestamp, l.timestamp)
+	}
+	l.blockHash = types.TruncateHash(upd.Hash)
+	l.blockNum = upd.Number
+	l.timestamp = timestamp
+	l.logIdx = ^uint32(0)
+	l.doneBlock = false
+	return nil
+}
+
+func (l *logContext) applyLog(parentBlock eth.BlockID, logIdx uint32, logHash types.TruncatedHash, execMsg *types.ExecutingMessage) error {
+	if err := l.inferFull(); err != nil {
+		return err
+	}
```

Comment:
rather than have `inferFull` be the responsibility of the new caller, why not have `inferFull` happen at the end of each of these functions?